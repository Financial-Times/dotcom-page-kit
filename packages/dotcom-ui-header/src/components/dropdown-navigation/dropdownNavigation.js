/* eslint-disable no-console */
import { PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST } from './constants'

/**
 * Enhance the dropdown menu for JS users. This will ensure we can
 * perform the same actions as we do for non-JS users, however it enables
 * more interactivity of the dropdown buttons and enable them to send
 * tracking events on click.
 */
const enhanceInteractivity = () => {
  // Use querySelectorAll as there could be multiple dropdowns on the page
  const dropdowns = document?.querySelectorAll('.o-header__dropdown')
  if (!dropdowns || dropdowns.length === 0) {
    return
  }

  dropdowns.forEach((dropdownContainer) => {
    const dropdownButton = dropdownContainer.querySelector('.o-header__dropdown-button')
    const closeDropdownButton = dropdownContainer.querySelector('.o-header__dropdown-close-button-mobile')

    // For Non-JS users the pointer events on the dropdown button are
    // disabled by default as it is the only way to enable the button
    // to both open and close the dropdown. This will not be needed for JS
    dropdownButton.style.pointerEvents = 'auto'
    dropdownButton.addEventListener('click', () => {
      // When the dropdown button has been clicked before and the dropdown is currently shown
      // reset the button's active data attribute and blur the button (to remove focus and hide the dropdown)
      if (dropdownButton.hasAttribute('data-dropdown-button-active')) {
        dropdownButton.removeAttribute('data-dropdown-button-active')
        // Unlike in no-js we want to have pointer events on the dropdown button
        // in order to send tracking events on click. However, to close the dropdown
        // we need to remove the focus from the button, so we blur it.
        dropdownButton.blur()
      } else {
        // When the dropdown has just been clicked,
        // so we know the dropdown is actively getting shown now
        dropdownButton.setAttribute('data-dropdown-button-active', 'true')
      }
    })

    // For Non-JS MOBILE users the closing of the dropdown
    // is handled by playing with the visibility and pointer events
    // of the close button which enables to remove the focus so we can close the dropdown,
    // this will not be needed for JS
    closeDropdownButton.style.visibility = 'visible'
    closeDropdownButton.style.pointerEvents = 'auto'
    // When the close button is clicked, reset the dropdown button so we can open the dropdown
    // correctly next time we click the dropdown button.
    // Also blur the close dropdown button (the mobile close button) to close the dropdown
    closeDropdownButton.addEventListener('click', () => {
      dropdownButton.removeAttribute('data-dropdown-button-active')
      closeDropdownButton.blur()
    })
  })

  // When clicking outside the dropdown, reset the dropdown button by
  // removing the data attribute from the dropdown button to enable the dropdown to be opened correctly again
  document.addEventListener('click', (event) => {
    dropdowns.forEach((dropdownContainer) => {
      const dropdownButton = dropdownContainer.querySelector('.o-header__dropdown-button')
      if (!dropdownContainer.contains(event.target)) {
        dropdownButton && dropdownButton.removeAttribute('data-dropdown-button-active')
      }
    })
  })

  // Observe when the sticky header becomes visible and when it hides
  // and when it does we need to close the dropdown if it is open
  // This is needed since there are two headers and each has a dropdown
  // so we shouldn't be showing the sticky header's dropdown when the
  // sticky header is not visible, or show the regular header's dropdown
  // when the sticky header is visible
  const stickyHeader = document?.querySelector('.o-header--sticky')
  if (!stickyHeader) {
    return
  }

  const stickyHeaderObserver = new IntersectionObserver(
    (changes) => {
      for (let change of changes) {
        if (
          document.documentElement.clientWidth >= 780 ||
          (document.documentElement.clientWidth < 780 && !change.isIntersecting)
        ) {
          const closestDropdownParent = document.activeElement.closest('.o-header__dropdown')
          if (closestDropdownParent) {
            const dropdownButton = closestDropdownParent.querySelector('.o-header__dropdown-button')
            dropdownButton.removeAttribute('data-dropdown-button-active')
            document.activeElement.blur()
          }
        }
      }
    },
    { threshold: [0.01] }
  )

  stickyHeaderObserver.observe(stickyHeader)
}

/**
 * Dispatches a custom event with tracking data when the dropdown content becomes visible.
 *
 * @param {Object} options - Configuration options for tracking.
 * @param {string} options.selector - The CSS selector used to identify the elements to track.
 * @param {string} options.intersectionObserverThreshold - Customise the how much of this element needs to be in the viewport to trigger the event.
 */
const trackDropdownView = (options) => {
  if (!window.IntersectionObserver) {
    return
  }

  const selector = options && options.selector
  const elementsToTrack = document.querySelectorAll(selector)
  if (elementsToTrack.length === 0) {
    return
  }

  const onChange = (changes, observer) => {
    changes.forEach((change) => {
      if (change.isIntersecting || change.intersectionRatio > 0) {
        const viewedEl = change.target
        const eventData = {
          action: 'view',
          category: 'component',
          className: viewedEl.className,
          componentContentId: viewedEl.dataset.id,
          url: window.document.location.href || null,
          nodeName: viewedEl.nodeName,
          href: viewedEl.href || false,
          text: viewedEl.text || false,
          role: viewedEl.role || false
        }

        document.body.dispatchEvent(new CustomEvent('oTracking.event', { detail: eventData, bubbles: true }))
        observer.unobserve(viewedEl)
      }
    })
  }

  const observer = new IntersectionObserver(onChange, {
    threshold: options?.intersectionObserverThreshold || [1.0]
  })

  elementsToTrack.forEach((el) => observer.observe(el))
}

/**
 * Dispatches a custom event with tracking data for Amplitude experiments.
 * 
 */
const trackDropdownExposure = () => {
  const flagProNavigation = document.querySelector('[data-flag-pro-navigation]')?.dataset.flagProNavigation

  if (flagProNavigation !== undefined) {
    const flagExposureValue = flagProNavigation === 'true' ? 'treatment' : 'control'

    document.body.dispatchEvent(
      new CustomEvent('oTracking.event', {
        detail: {
          category: 'amplitudeExperiment',
          action: 'exposure',
          event_properties: {
            flag_key: 'pro-navigation',
            variant: flagExposureValue,
            experiment_key: 'exp-1'
          }
        },
        bubbles: true
      })
    )
  }
}

/**
 * Updates the links in the Pro Navigation dropdown.
 *
 * @param {Object} options - Configuration options for updating the dropdown links.
 * @param {string} options.selector - The CSS class selector used to identify the dropdown(s) to update. This property is required when initializing the dropdown component
 * @param {string} options.trackingKey - A key used for tracking user interactions with the dropdown links and toggler.
 * @param {Array} options.defaultLinks - The default list of links to display in the dropdown if the API call fails or returns no data.
 * @param {string} options.proNavigationApi - The URL of the API endpoint to fetch the updated links.
 */
const updateProNavigationLinks = async (options) => {
  const { selector, trackingKey, defaultLinks, proNavigationApi } = options

  const proDropdowns = document.querySelectorAll(`.o-header__dropdown.${selector}`)
  if (proDropdowns.length === 0) {
    return
  }

  try {
    const links = await fetchLinks(proNavigationApi)

    if (
      !links ||
      !Array.isArray(links) ||
      links.length === 0 ||
      JSON.stringify(defaultLinks) === JSON.stringify(links)
    ) {
      return
    }

    proDropdowns.forEach((dropdown) => updateLinksList(dropdown, links, trackingKey))
  } catch (error) {
    const isFetchError = error.message.includes('fetch')
    const eventData = {
      action: isFetchError ? 'fetch' : 'update',
      category: 'error',
      component_name: 'dropdown-navigation',
      errorMessage: error.message
    }
    document.body.dispatchEvent(new CustomEvent('oTracking.event', { detail: eventData, bubbles: true }))
  }
}

const fetchLinks = async (url) => {
  const response = await fetch(url, { credentials: 'include' })
  if (!response.ok) {
    throw new Error(`Error during navigation links fetch! Status: ${response.status}`)
  }
  return response.json()
}

const updateLinksList = (dropdownEl, links, trackingKey) => {
  const list = dropdownEl.querySelector('.o-header__dropdown-list')
  if (!list) {
    return
  }

  // Determine the list item template
  const label = dropdownEl.querySelector('.o-header__dropdown-list-item-label-container')
  const listItem = label?.closest('.o-header__dropdown-list-item') || list.firstElementChild

  if (!listItem) {
    return
  }

  // Build the updated list
  const updatedListFragment = document.createDocumentFragment()
  links.forEach((link) => {
    const updatedItem = buildListItem(listItem, label, link, trackingKey)
    updatedItem && updatedListFragment.append(updatedItem)
  })

  // Update the DOM
  if (updatedListFragment.childNodes.length > 0) {
    list.replaceChildren(updatedListFragment)
  }
}

const buildListItem = (listItem, label, link, trackingKey) => {
  const listItemClone = listItem.cloneNode(true)
  const a = listItemClone.querySelector('a')
  const icon = a.querySelector('.o-header__dropdown-icon')
  const span = a.querySelector('span:not([class])')

  if (!listItemClone || !a || !icon || !span) {
    return
  }

  // Update class name
  listItemClone.classList.toggle('o-header__dropdown-list-divider', link.hasBottomLine)

  // Update link attributes
  a.setAttribute('href', link.href)
  const availableTrackingKey = trackingKey || a.dataset.trackingKey
  if (availableTrackingKey) {
    a.setAttribute('data-trackable', `${availableTrackingKey}_${link.id}_clicked`)
  }

  // Update icon and text
  icon.classList.forEach(
    (c) =>
      /^(?!.*__)[\w-]+-icon$/.test(c) &&
      icon.classList.replace(c, `${link.hasAccess ? link.icon : 'lock'}-icon`)
  )
  span.innerText = link.title

  // Update label: If a label is present, the cloned li item will need tidying up if no label is required
  if (label && !link.hasLabel) {
    a.lastElementChild?.className?.includes('label-container') && a.lastElementChild.remove()
  }

  return listItemClone
}

const init = () => {
  enhanceInteractivity()

  trackDropdownView({ selector: '.o-header__dropdown-content', intersectionObserverThreshold: 0.8 })
  trackDropdownExposure()

  updateProNavigationLinks({
    selector: 'pro_navigation',
    trackingKey: 'pro_navigation',
    defaultLinks: PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST,
    proNavigationApi: 'https://pro-navigation.ft.com/api/links'
  })

  // Add any dropdown navigation-specific initializations below.
}

export const DropdownNavigation = {
  init
}
