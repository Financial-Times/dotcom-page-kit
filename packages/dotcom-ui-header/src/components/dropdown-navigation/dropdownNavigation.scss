// Start Dropdown styles
@import '@financial-times/o3-foundation/css/professional.css';

%icon-base {
  /* Create a square the size we want an icon */
  display: inline-block;
  width: var(--o3-spacing-s);
  height: var(--o3-spacing-s);
  /* Set the icon colour, In this case match the current foreground text colour. */
  background-color: currentColor;
  /* Mask the square with an Origami icon. */
  mask-repeat: no-repeat;
  mask-size: contain;
}

@mixin hidden {
  opacity: 0;
  visibility: hidden;
}

@mixin visible {
  opacity: 1;
  visibility: visible;
}

@mixin delay-to-propagate-click-events {
  transition: opacity 0.3s ease, visibility 0.3s ease;
}

@mixin disable-setting-focus {
  pointer-events: none;
}

@mixin enable-setting-focus {
  pointer-events: auto;
}

.o-header__dropdown {
  position: relative;
  z-index: 100;
}

.o-header__dropdown-button {
  display: flex;
  height: var(--o3-spacing-l);
  gap: var(--o3-spacing-4xs);
  padding: 0;
  margin: 0;
  background: none;
  border: none;
  align-items: center;
  justify-content: center;

  .o-header__dropdown-button-user-icon-wrapper {
    height: var(--o3-spacing-l);
    width: var(--o3-spacing-l);
    background: linear-gradient(153deg, #262a33 54.99%, #434958 99.56%);
    border-radius: var(--o3-spacing-m);
    border: 2px solid #262a33;
    align-content: center;

    .o-header__dropdown-icon {
      background-color: #c0efd8;
    }
  }

  .chevron-icon {
    background-color: var(--o3-color-palette-black);
    mask-image: var(--o3-icon-chevron-down);
    display: none;
  }
}

.o-header__dropdown:focus-within > .o-header__dropdown-button {
  // This is needed for No-JS as we need to be able to close the dropdown
  // when we click the dropdown button again
  // This will remove the focus from the button and in turn deactivate the dropdown conent :foucus-within pseudo selector which will hide the dropdown.
  @include disable-setting-focus;

  .chevron-icon {
    mask-image: var(--o3-icon-chevron-up);
  }
}

.o-header__dropdown-content {
  position: fixed;
  // Move the dropdown off the viewport so the component view tracking
  // doesn't trigger while the dropdown is still hidden
  top: -100vh;
  left: 100vw;
  /* Start offscreen */
  width: 0;
  height: 0;
  background: var(--o3-color-palette-white);

  /* Hide dropdown by default */
  @include disable-setting-focus;
  @include hidden;
  @include delay-to-propagate-click-events;
}

/* Show dropdown when the button OR dropdown itself is focused */
.o-header__dropdown:focus-within > .o-header__dropdown-content {
  @include visible;
  // Ensure that clicking inside the dropdown does not close it
  @include enable-setting-focus;
  // Move into the viewport (the horizontal position is handled separately
  // for desktop and mobile, where for mobile it is done by using animations)
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
}

.o-header__dropdown-header {
  display: flex;
  justify-content: space-between;
  padding: var(--o3-spacing-2xs);
  background: linear-gradient(153deg, #262a33 54.99%, #434958 99.56%);
}

.o-header__dropdown-title {
  font-size: var(--o3-type-label-font-size);
  line-height: var(--o3-type-label-line-height);
  font-weight: var(--o3-type-label-font-weight);
  font-family: var(--o3-type-label-font-family);
  text-transform: var(--o3-type-label-text-case);
  padding: var(--o3-spacing-5xs) 0;
  color: var(--o3-color-palette-mint);
}

.o-header__dropdown-list {
  list-style: none;
  margin: 0;
  padding: 0;
  border-bottom-left-radius: var(--o3-spacing-5xs);
  border-bottom-right-radius: var(--o3-spacing-5xs);
  border: none;
  max-height: 100%;
  overflow-y: scroll;
  overscroll-behavior: none;
}

.o-header__dropdown-list-item {
  list-style: none;

  &:hover {
    background-color: #3333330d;
  }

  &:active {
    background-color: #3333331a;
  }
}

.o-header__dropdown-list-item-link {
  display: flex;
  align-items: center;
  gap: var(--o3-spacing-4xs);
  padding: var(--o3-spacing-2xs);
  text-decoration: none;
  font-size: var(--o3-type-body-base-font-size);
  line-height: var(--o3-type-body-base-line-height);
  font-weight: var(--o3-type-body-base-font-weight);
  font-family: var(--o3-type-body-base-font-family);
  color: var(--o3-color-palette-black);

  &:hover {
    color: var(--o3-color-palette-black);
  }

  &:hover > .o-header__dropdown-list-item-details-container {
    text-decoration: underline;
    text-underline-offset: var(--o3-spacing-5xs);
  }
}

.o-header__dropdown-list-item-details-container {
  display: flex;
  gap: var(--o3-spacing-2xs);
}

.o-header__dropdown-list-divider {
  border-bottom: 1px solid #d6d6d6;
}

.o-header__dropdown-icon {
  @extend %icon-base;
}

@each $icon-name
  in (dashboard, bookmark, lock, list, radar, highlight, share, mail, settings, exit, cross, user, users)
{
  .#{$icon-name}-icon {
    mask-image: var(--o3-icon-#{$icon-name});
  }
}

.o-header__dropdown-list-pro-label {
  align-items: center;
  border-radius: 2px;
  padding: var(--o3-spacing-5xs) 10px;
  background-color: var(--o3-color-use-case-link-underline-hover);
  font-size: var(--o3-type-label-font-size);
  line-height: var(--o3-type-label-line-height);
  font-weight: var(--o3-type-label-font-weight);
  font-family: var(--o3-type-label-font-family);
  text-transform: var(--o3-type-label-text-case);
  color: var(--o3-color-palette-slate);
}

.o-header__dropdown-close-button-mobile {
  @extend .o-header__dropdown-icon;
  @extend .cross-icon;
  background-color: var(--o3-color-palette-white);
}

/* Close the menu when clicking the close button */
.o-header__dropdown-close-button-mobile:focus {
  // Applying visibility:none to the close icon will remove the focus from it and in turn deactivate the dropdown conent :foucus-within pseudo selector which will hide the dropdown.
  @include hidden;
  @include disable-setting-focus;
}

@media (max-width: 780px) {
  // Animations for sliding the dropdown in and out
  // on mobile devices
  @keyframes slideIn {
    0% {
      left: 100vw;
    }

    100% {
      left: 0;
    }
  }

  @keyframes slideOut {
    0% {
      left: 0;
    }

    100% {
      left: 100vw;
    }
  }

  // Slide the dropdown in
  .o-header__dropdown:focus-within > .o-header__dropdown-content {
    animation: slideIn 0.5s ease-in-out forwards;
    animation-duration: 0.5s !important;
  }

  // Slide the dropdown out
  .o-header__dropdown:not(:focus-within) > .o-header__dropdown-content {
    animation: slideOut 0.5s ease-in-out forwards;
    animation-duration: 0.5s !important;

    // Delay the actual removal of the dropdown from the viewport (and its visibility)
    // Until the slide out animation is complete
    transition: visibility 0s linear 0.5s, opacity 0s linear 0.5s, width 0s linear 0.5s, height 0s linear 0.5s,
      top 0s linear 0.5s;
    /* Delay hiding to match animation duration */
  }
}

@media (min-width: 780px) {
  .o-header__dropdown-button {
    .chevron-icon {
      display: block;
    }
  }

  .o-header__dropdown-content {
    top: -100vh;
    left: -100vw;
    width: auto;
    height: auto;
    margin-top: var(--o3-spacing-2xs);
    border-radius: var(--o3-spacing-5xs);
    box-shadow: 2px 2px 4px 0px rgba(0, 0, 0, 0.05);
    overflow: hidden; // for the border-radius to work
  }

  .o-header__dropdown:focus-within > .o-header__dropdown-content {
    @include visible;
    // Ensure that clicking inside the dropdown does not close it
    @include enable-setting-focus;
    // Move into the viewport
    position: absolute;
    right: 0px;
    width: auto;
    height: auto;
    top: unset;
    left: unset;
  }

  .o-header__dropdown-close-button-mobile {
    display: none;
  }

  .o-header__dropdown-list {
    max-height: 550px;
    border: 1px solid #d6d6d6;
    border-top: none;
  }
}
