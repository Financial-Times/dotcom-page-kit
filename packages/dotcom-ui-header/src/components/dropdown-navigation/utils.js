export const isEqual = (obj1, obj2) => {
  if (obj1 === obj2) {
    return true
  }

  if (typeof obj1 !== 'object' || typeof obj2 !== 'object' || obj1 === null || obj2 === null) {
    return false
  }

  const keys1 = Object.keys(obj1)
  const keys2 = Object.keys(obj2)

  if (keys1.length !== keys2.length) {
    return false
  }

  for (let i = 0; i < keys1.length; i++) {
    const k1 = keys1[i]
    const k2 = keys2[i]
    if (k1 !== k2) {
      return false
    }

    if (!isEqual(obj1[k1], obj2[k2])) {
      return false
    }
  }

  return true
}
