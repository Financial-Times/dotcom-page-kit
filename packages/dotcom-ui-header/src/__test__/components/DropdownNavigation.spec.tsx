/**
 * @jest-environment jsdom
 */
import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { DropdownNavigation } from '../../components/dropdown-navigation/dropdownNavigation'
import {PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST} from '../../components/dropdown-navigation/dropdownNavigationDefaultLists'

describe('DropdownNavigation', () => {
  it('renders correctly', () => {
    const { baseElement: dropdown } = render(
      <DropdownNavigation
        options={PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST}
        headerContent="Test Title"
        trackingKey="test_tracking"
      />
    )

    expect(dropdown).toMatchSnapshot()
  })

  it('renders label', () => {
    const MockLabel = () => <div>Mock Label</div>
    const mockLabelCount = PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST.filter((item) => item.hasLabel).length
    const { baseElement: dropdown, queryAllByText } = render(
      <DropdownNavigation
        options={PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST}
        headerContent="Test Title"
        trackingKey="test_tracking"
        label={<MockLabel />}
      />
    )

    const label = queryAllByText('Mock Label')

    expect(dropdown).toMatchSnapshot()
    expect(label).toHaveLength(mockLabelCount)
  })

  it('renders custom header', () => {
    const MockHeader = () => <div>Dropdown Custom Header</div>
    const { getByText } = render(
      <DropdownNavigation
        options={PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST}
        headerContent={<MockHeader />}
        trackingKey="test_tracking"
      />
    )

    const header = getByText('Dropdown Custom Header')

    expect(header).toBeInTheDocument()
  })

  it('sets tracking attributes correctly', () => {
    const trackingKey = 'test_tracking'
    const { getAllByRole, baseElement } = render(
      <DropdownNavigation
        options={PRO_NAVIGATION_DROPDOWN_DEFAULT_LIST}
        headerContent="Test Title"
        trackingKey={trackingKey}
      />
    )

    const toggleButton = getAllByRole('button')[0]
    const links = getAllByRole('link')
    const contentContainer = baseElement.querySelector('[data-id="dropdown-content"]')!

    expect(toggleButton.getAttribute('data-trackable')).toContain(trackingKey)
    links.forEach((link) => {
      expect(link.getAttribute('data-trackable')).toContain(trackingKey)
    })
    expect(contentContainer).toBeInTheDocument()
    expect(contentContainer.getAttribute('data-o-tracking-view')).toContain(trackingKey)
  })
})
