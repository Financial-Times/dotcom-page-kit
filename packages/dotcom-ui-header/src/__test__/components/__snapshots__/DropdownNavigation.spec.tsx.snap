// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownNavigation renders correctly 1`] = `
<body>
  <div>
    <nav
      aria-describedby="dropdown-title"
      class="o-header__dropdown"
    >
      <button
        aria-controls="dropdown-options"
        aria-label="Dropdown menu has opened on focus, press Tab to access links."
        class="o-header__dropdown-button"
        data-trackable="test_tracking_toggle_click"
        tabindex="0"
      >
        <div
          class="o-header__dropdown-button-user-icon-wrapper"
        >
          <span
            aria-hidden="true"
            class="o-header__dropdown-icon user-icon"
          />
        </div>
        <span
          aria-hidden="true"
          class="o-header__dropdown-icon chevron-icon"
        />
      </button>
      <div
        class="o-header__dropdown-content"
        data-id="dropdown-content"
        data-o-tracking-view="test_tracking_component_view"
        id="dropdown-options"
        role="group"
        tabindex="-1"
      >
        <div
          class="o-header__dropdown-header"
        >
          <span
            class="o-header__dropdown-title"
            id="dropdown-title"
          >
            Test Title
          </span>
          <button
            aria-label="Close dropdown menu"
            class="o-header__dropdown-close-button-mobile"
            tabindex="0"
          />
        </div>
        <ul
          class="o-header__dropdown-list"
        >
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-workspace_clicked"
              href="https://workspace.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon dashboard-icon"
                />
                <span>
                  Workspace
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-feed_clicked"
              href="https://www.ft.com/myft/following/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon list-icon"
                />
                <span>
                  Feed
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item o-header__dropdown-list-divider"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-mpr_clicked"
              href="https://professional-monetary-policy-radar.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon radar-icon"
                />
                <span>
                  Monetary Policy Radar
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-saved-articles_clicked"
              href="https://www.ft.com/myft/saved-articles/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon bookmark-icon"
                />
                <span>
                  Saved articles
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-highlights_clicked"
              href="https://highlights-dashboard.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon highlight-icon"
                />
                <span>
                  Highlights
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item o-header__dropdown-list-divider"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-advanced-sharing_clicked"
              href="https://sharing-dashboard.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon share-icon"
                />
                <span>
                  Advanced Sharing
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-newsletters_clicked"
              href="https://www.ft.com/newsletters"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon mail-icon"
                />
                <span>
                  Newsletters
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-account-settings_clicked"
              href="https://www.ft.com/myaccount/personal-details"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon settings-icon"
                />
                <span>
                  Account settings
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-sign-out_clicked"
              href="https://www.ft.com/logout"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon exit-icon"
                />
                <span>
                  Sign out
                </span>
              </div>
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </div>
</body>
`;

exports[`DropdownNavigation renders label 1`] = `
<body>
  <div>
    <nav
      aria-describedby="dropdown-title"
      class="o-header__dropdown"
    >
      <button
        aria-controls="dropdown-options"
        aria-label="Dropdown menu has opened on focus, press Tab to access links."
        class="o-header__dropdown-button"
        data-trackable="test_tracking_toggle_click"
        tabindex="0"
      >
        <div
          class="o-header__dropdown-button-user-icon-wrapper"
        >
          <span
            aria-hidden="true"
            class="o-header__dropdown-icon user-icon"
          />
        </div>
        <span
          aria-hidden="true"
          class="o-header__dropdown-icon chevron-icon"
        />
      </button>
      <div
        class="o-header__dropdown-content"
        data-id="dropdown-content"
        data-o-tracking-view="test_tracking_component_view"
        id="dropdown-options"
        role="group"
        tabindex="-1"
      >
        <div
          class="o-header__dropdown-header"
        >
          <span
            class="o-header__dropdown-title"
            id="dropdown-title"
          >
            Test Title
          </span>
          <button
            aria-label="Close dropdown menu"
            class="o-header__dropdown-close-button-mobile"
            tabindex="0"
          />
        </div>
        <ul
          class="o-header__dropdown-list"
        >
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-workspace_clicked"
              href="https://workspace.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon dashboard-icon"
                />
                <span>
                  Workspace
                </span>
              </div>
              <div>
                Mock Label
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-feed_clicked"
              href="https://www.ft.com/myft/following/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon list-icon"
                />
                <span>
                  Feed
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item o-header__dropdown-list-divider"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-mpr_clicked"
              href="https://professional-monetary-policy-radar.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon radar-icon"
                />
                <span>
                  Monetary Policy Radar
                </span>
              </div>
              <div>
                Mock Label
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-saved-articles_clicked"
              href="https://www.ft.com/myft/saved-articles/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon bookmark-icon"
                />
                <span>
                  Saved articles
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-highlights_clicked"
              href="https://highlights-dashboard.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon highlight-icon"
                />
                <span>
                  Highlights
                </span>
              </div>
              <div>
                Mock Label
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item o-header__dropdown-list-divider"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-advanced-sharing_clicked"
              href="https://sharing-dashboard.ft.com/"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon share-icon"
                />
                <span>
                  Advanced Sharing
                </span>
              </div>
              <div>
                Mock Label
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-newsletters_clicked"
              href="https://www.ft.com/newsletters"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon mail-icon"
                />
                <span>
                  Newsletters
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-account-settings_clicked"
              href="https://www.ft.com/myaccount/personal-details"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon settings-icon"
                />
                <span>
                  Account settings
                </span>
              </div>
            </a>
          </li>
          <li
            class="o-header__dropdown-list-item false"
          >
            <a
              class="o-header__dropdown-list-item-link"
              data-trackable="test_tracking_link-sign-out_clicked"
              href="https://www.ft.com/logout"
            >
              <div
                class="o-header__dropdown-list-item-details-container"
              >
                <span
                  aria-hidden="true"
                  class="o-header__dropdown-icon exit-icon"
                />
                <span>
                  Sign out
                </span>
              </div>
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </div>
</body>
`;
